Do not allow SETTINGS after FORMAT for INSERT queries

azat opened the pull request:
Changelog category (leave one):
- Backward Incompatible Change

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Do not allow SETTINGS after FORMAT for INSERT queries (there is compatibility setting `parser_settings_after_format_compact` to accept such queries, but it is turned OFF by default)

Parsing SETTINGS after FORMAT, that has been introduced in [1], can
interpret SETTING as some values, which is misleading.

  [1]: https://github.com/ClickHouse/ClickHouse/pull/4174/files#diff-ba7bd0657630b1cd94cf6ed364bd857338096f49f66dc82918438d6745753775R106

Note, that we are touching only INSERT queries, not SELECT, since this
is a backward incompatible change, and in case of modifying SELECT it
can break too much.

Fixes: #35100
Fixes: #20343
Cc: @alexey-milovidov

azat committed:
Do not allow SETTINGS after FORMAT for INSERT queries

Parsing SETTINGS after FORMAT, that has been introduced in [1], can
interpret SETTING as some values, which is misleading.

  [1]: https://github.com/ClickHouse/ClickHouse/pull/4174/files#diff-ba7bd0657630b1cd94cf6ed364bd857338096f49f66dc82918438d6745753775R106

Note, that we are touching only INSERT queries, not SELECT, since this
is a backward incompatible change, and in case of modifying SELECT it
can break too much.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>
Fixes: #35100
Fixes: #20343

azat commented:
>[Style Check (actions) — Check style failed](https://s3.amazonaws.com/clickhouse-test-reports/35883/e6ba60d2422126250fef32ec24abd54fd6f2f15f/style_check__actions_/style_output.txt)

- Likely broken in #35858

vitlibar commented on src/Parsers/ParserInsertQuery.cpp:
```diff
@@ -133,6 +133,20 @@ bool ParserInsertQuery::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
+        /// - INSERT WATCH
+        before_values = pos;
+    }
+
```
We can write `Pos before_values = pos;` here to assign `before_values` in only one place

azat replied:
Right, thanks.

vitlibar commented on src/Parsers/ASTInsertQuery.cpp:
```diff
@@ -86,33 +86,41 @@ void ASTInsertQuery::formatImpl(const FormatSettings & settings, FormatState & s
+    {
+        settings.ostr << (settings.hilite ? hilite_keyword : "") << settings.nl_or_ws << "SETTINGS " << (settings.hilite ? hilite_none : "");
+        settings_ast->formatImpl(settings, state, frame);
+        settings.ostr << settings.nl_or_ws;
```
It seems if we replace this line with

```
space = settings.nl_or_ws
```

then the conditions below 
```
        if (!settings_ast)
            settings.ostr << ' ';
```

will turn into just `settings.ostr << space`

azat replied:
This will break compatibility for INSERT w/o SETTINGS formatted in oneline, but I got the idea, will simplify the code.

azat committed:
Introduce compatiblity mode for SETTINGS after FORMAT

Add allow_settings_after_format_in_insert setting, OFF by default.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>
v2: s/parser_settings_after_format_compact/allow_settings_after_format_in_insert/ (suggested by vitlibar)
v3: replace ParserSettings with a flag (requested by vitlibar)

azat commented:
>[Stateless tests flaky check (address, actions) — Timeout, fail: 0, passed: 2161](https://s3.amazonaws.com/clickhouse-test-reports/35883/261106544d1c26330f56310983f92037d545fdb4/stateless_tests_flaky_check__address__actions_.html)

Timeout - OK, no failures

vitlibar commented on src/Parsers/ParserQuery.h:
```diff
@@ -10,12 +10,21 @@ class ParserQuery : public IParserBase
+    explicit ParserQuery(const char * end_, const ParserSettings & parser_settings_)
+        : end(end_)
+        , parser_settings(parser_settings_)
+    {}
```
Looks very strange, why do you need to change so common classes?
I thought you were going to just check your new setting `parse_settings_after_format` in `ParserInsertQuery` and that's all

azat replied:
But how would you pass this setting to `ParserInsertQuery`?
Entrypoint for the parsing is - `ParserQuery`

vitlibar replied:
Probably it's better not to use the structure and just pass the parameter to constructors?

vitlibar replied:
Probably it's better not to use the structure and just pass the parameter to constructors?

azat replied:
I would prefer to keep structure, it is always easier to extend structure then arguments.
But if you have strong objections, then OK, I can replace it with the variable.

vitlibar commented on src/Parsers/ParserQuery.h:
```diff
@@ -10,12 +10,21 @@ class ParserQuery : public IParserBase
+    explicit ParserQuery(const char * end_)
+        : end(end_)
+    {}
+    /// For production parser.
```
We can join both those versions to a single constructor

```
explicit ParserQuery(const char * end_, const ParserSettings & parser_settings_ = {})
```

vitlibar commented on src/Parsers/IParser.h:
```diff
@@ -49,6 +49,13 @@ struct Expected
+  */
+struct ParserSettings
+{
+    bool parser_settings_after_format_compact = false;
```
It seems the name `allow_settings_after_format_in_insert` is shorter and clearer.

azat replied:
Yep, thanks, will rename.
